clang/lib/CIR/CodeGen/CIRGenBuiltinAArch64.cpp

```diff
@@ -1569,6 +1569,44 @@ mlir::Value CIRGenFunction::buildScalarOrConstFoldImmArg(unsigned ICEArguments,
   return Arg;
 }
 
+static mlir::Value buildArmLdrexNon128Intrinsic(unsigned int builtinID,
+                                                const CallExpr *clangCallExpr,
+                                                CIRGenFunction &cgf) {
+  StringRef intrinsicName;
+  if (builtinID == clang::AArch64::BI__builtin_arm_ldrex) {
+    intrinsicName = "llvm.aarch64.ldxr";
+  } else {
+    llvm_unreachable("Unknown builtinID");
+  }
+  // Argument
+  mlir::Value loadAddr = cgf.buildScalarExpr(clangCallExpr->getArg(0));
+  // Get Instrinc call
+  CIRGenBuilderTy &builder = cgf.getBuilder();
+  QualType clangResTy = clangCallExpr->getType();
+  mlir::Type realResTy = cgf.ConvertType(clangResTy);
+  // Return type of LLVM intrinsic is defined in Intrinsic<arch_type>.td,
+  // which can be found under LLVM IR directory.
+  mlir::Type funcResTy = builder.getSInt64Ty();
+  mlir::Location loc = cgf.getLoc(clangCallExpr->getExprLoc());
+  mlir::cir::IntrinsicCallOp op = builder.create<mlir::cir::IntrinsicCallOp>(
+      loc, builder.getStringAttr(intrinsicName), funcResTy, loadAddr);
+  mlir::Value res = op.getResult();
+
+  // Convert result type to the expected type.
+  if (mlir::isa<mlir::cir::PointerType>(realResTy)) {
+    return builder.createIntToPtr(res, realResTy);
+  }
+  mlir::cir::IntType intResTy =
```

Comment:
Thanks for thinking about the case where realResTy isn't a int type, but can that really happen here? Let's just mimic OG closer and nest the bitcasts directly into the result (no need for the check).

Reply:
this logic is actually from OG, and the OG test case you referred has the case for this realResTy not int type, 
https://github.com/llvm/clangir/blob/cef345930be5fce6707a2fdb22ee7119f04c4221/clang/test/CodeGen/builtins-arm-exclusive.c#L46

Reply:
I know, what I'm talking about is:
```
mlir::cir::IntType intResTy = builder.getSIntNTy(cgf.CGM.getDataLayout().getTypeSizeInBits(realResTy));
if (mlir::isa<mlir::cir::IntType>(realResTy)) {
    return builder.createIntCast(intCastRes, realResTy);
  } else {
    // Above cases should cover most situations and we have test coverage.
    llvm_unreachable("Unsupported return type for now");
  }
```
Looking at OG, looks like it should be instead:
```
mlir::cir::IntType intResTy = builder.getSIntNTy(cgf.CGM.getDataLayout().getTypeSizeInBits(realResTy));
return builder.createIntCast(builder.createIntCast(res, intResTy), realResTy);
```
I don't see anything in OG checking if the resulting type is an int, so I don't understand why you are adding that here (it's nice to be worried about extra unhandled cases, but I don't think it can happen here).

Reply:
OG uses llvm IRBuilder's CreateBitCast helper to create the cast, which doesn't have to specify CastKind as CIR requires, that's why need to I know if realResType is int type.
If its float type, I need to pick a different CastKind.

Reply:
Gotcha, good catch! Thanks!